File: run-ssh/go/bash_cmd_REST_API.go
package main

import (
    "fmt"
    "net"
    "net/http"
    "os/exec"
    "strconv"
    "strings"

    "github.com/gorilla/mux"
)

func getAvailablePort(w http.ResponseWriter, r *http.Request) {
    vars := mux.Vars(r)
    ip := vars["ip"]
    start, _ := strconv.Atoi(vars["start"])
    end, _ := strconv.Atoi(vars["end"])

    for port := start; port <= end; port++ {
        address := fmt.Sprintf("%s:%d", ip, port)
        listener, err := net.Listen("tcp", address)
        if err == nil {
            listener.Close()
            fmt.Fprintf(w, `{"port": %d}`, port)
            return
        }
    }
    fmt.Fprintf(w, `{"error": "No available port found"}`)
}

func runCommand(w http.ResponseWriter, r *http.Request) {
    command := r.FormValue("command")
    cmd := exec.Command("sh", "-c", command)
    output, err := cmd.CombinedOutput()
    if err != nil {
        fmt.Fprintf(w, `{"error": "%s"}`, err)
        return
    }
    fmt.Fprintf(w, `{"output": "%s"}`, strings.TrimSuffix(string(output), "\n"))
}

func main() {
    r := mux.NewRouter()
    r.HandleFunc("/get_port/{ip}/{start:[0-9]+}/{end:[0-9]+}", getAvailablePort).Methods("GET")
    r.HandleFunc("/run", runCommand).Methods("POST")
    http.ListenAndServe(":8888", r)
}

File: run-ssh/bash_cmd_REST_API.py
from flask import Flask, request, jsonify
import subprocess
import socket

app = Flask(__name__)

@app.route('/get_port/<ip>/<int:start>/<int:end>', methods=['GET'])
def get_available_port(ip, start, end):
    s = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
    for port in range(start, end):
        try:
            s.bind((ip, port))
            s.close()
            return jsonify({'port': port})
        except OSError:
            continue
    return jsonify({'error': 'No available port found'})


@app.route('/run', methods=['POST'])
def run_command():
    command = request.json['command']
    process = subprocess.Popen(command, stdout=subprocess.PIPE, shell=True)
    # run the process with exit code
    output, error = process.communicate()
    exit_code = process.wait()

    return {
        'output': output.decode('utf-8'),
        'error': error,
        'exit_code': exit_code
    }

if __name__ == '__main__':
    # run app on 0.0.0.0:8888
    app.run(host='0.0.0.0', port=8888, debug=True)